src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(ForwardCancellationTokenToInvocationsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingCancellationToken, out INamedTypeSymbol? cancellationTokenType))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                if (!(context.ContainingSymbol is IMethodSymbol containingSymbol))
+                {
+                    return;
+                }
+
+                if (!ShouldDiagnose(
+                    invocation,
+                    containingSymbol,
+                    cancellationTokenType,
+                    out string? cancellationTokenParameterName))
+                {
+                    return;
+                }
+
+                // Only underline the method name, not the whole invocation
+                SyntaxNode? expressionNode = GetMethodNameNode(context.Operation.Syntax);
+                if (expressionNode != null)
+                {
+                    context.ReportDiagnostic(expressionNode.CreateDiagnostic(ForwardCancellationTokenToInvocationsRule, cancellationTokenParameterName, invocation.TargetMethod.Name));
+                }
+            },
+            OperationKind.Invocation);
+        }
+
+        // Determines if an invocation should trigger a diagnostic for this rule or not.
+        private static bool ShouldDiagnose(
+            IInvocationOperation invocation,
+            IMethodSymbol containingSymbol,
+            INamedTypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            IMethodSymbol method = invocation.TargetMethod;
+
+            // Verify that the current invocation is not passing a token already
+            if (invocation.Arguments.Any(a => a.Parameter.Type.Equals(cancellationTokenType) && !a.IsImplicit))
```

Comment:
This is a minor suggestion, but wouldn't it be faster/better to have a lookup starting from the end as we know that most methods following the guidelines will have the token at/near the end? It can hurt readability for a small perf gain though.

Reply:
Good point. Done. I had two places that were benefited with this.